Feat#31. 게시판 리스트 조회 API

JinDDung2 opened the pull request:
### Issue
 
- close #31

### Summary
- 태그 별로(여러개 태그 가능) 게시판 리스트 조회
- 태그가 존재하지 않으면 전체 태그로 간주하고 게시판 리스트 조회

### Description
- 게시판 리스트 조회 request에 태그 리스트 타입을 Long에서 Int로 바꿨습니다. Tag 엔티티는 Int로 되어있기 때문입니다.
- CursorPage를 커스터마이징 했습니다
- 응답에 다음 커서로 이동하기 위한 url을 등록했습니다.
- 만약 마지막 페이지라면 다음 커서로 이동하기 위한 url은 null로 되어 응답값에 포함되지 않습니다.

JinDDung2 committed:
✨ 게시판 리스트 조회 기능

RokwonK commented on adevspoon-api/src/main/kotlin/com/adevspoon/api/board/dto/request/BoardListRequest.kt:
```diff
@@ -5,13 +5,13 @@ import jakarta.validation.constraints.Positive
-    val tag: List<Long>,
+    val tag: List<Int>,
     @Schema(description = "가져올 게시글 갯수", nullable = true, defaultValue = "10")
     @field:Positive(message = "0보다 큰 수를 입력해주세요")
-    val take: Long = 10,
+    val take: Int = 10,
```
tag id는 괜찮아도 BoardPost의 경우 id가 Long타입 입니다. 함부로 Int로 바꾸면 문제가 생겨요!

JinDDung2 replied:
take의 역할이 pageSize 역할이라 생각했어요. 그래서 데이터를 가져올 때 몇 개 까지 가져올까? 라는 생각으로 접근하니 Int로 처리하게 되었어요. id를 다루는 부분은 startId라고 생각해서 그 부분을 시작점으로 pageable에 넘겼어요.
제가 이해한 take의 역할이 다를까요?

RokwonK commented on adevspoon-api/src/main/kotlin/com/adevspoon/api/board/dto/request/BoardListRequest.kt:
```diff
@@ -5,13 +5,13 @@ import jakarta.validation.constraints.Positive
     @Schema(description = "가져올 게시글 시작 Id(이 Id보다 작은 Id의 게시글들을 가져옴), null이면 가장 최신데이터부턱 가져옴", nullable = true, defaultValue = "0")
     @field:Positive(message = "0보다 큰 수를 입력해주세요")
     val startId: Long? = null,
     @Schema(description = "유저 Id(게시글 작성자 Id)로 필터링", nullable = true, defaultValue = "null")
-    val userId: Long?,
+    val targetUserId: Long?,
 )
```
배포된 api에서 userId 라는 key로 쿼리를 사용하고 있기때문에 바꾸면 큰일납니다. 정 targetUserId라는 이름으로 사용하고자 한다면 @JsonProperty 사용해주세요.

JinDDung2 replied:
userId로 고치는게 나을거 같아요.

RokwonK commented on adevspoon-domain/src/main/kotlin/com/adevspoon/domain/board/service/BoardPostDomainService.kt:
```diff
@@ -28,12 +33,46 @@ class BoardPostDomainService(
         val boardPost = boardPostRepository.findByIdOrNull(postId) ?: throw BoardPostNotFoundException()
         boardPost.increaseViewCount()
         boardPostRepository.save(boardPost)
+
+        val userId = boardPost.user?.id
+        userId?.let { BoardPost.from(boardPost, memberDomainService.getMemberProfile(it)) } ?: throw MemberNotFoundException()
+
```
이 로직은 무엇 때문에 존재하는 건가요?

JinDDung2 replied:
이전 코드가 로그인한 유저의 데이터를 가져오고 있었어요. 그래서 board를 작성한 유저를 가져오기 위한 작업으로 코드를 추가했어요.

RokwonK commented on adevspoon-domain/src/main/kotlin/com/adevspoon/domain/board/service/BoardPostDomainService.kt:
```diff
@@ -28,12 +33,46 @@ class BoardPostDomainService(
     }
 
+    fun getBoardPostsByTags(tags: List<Int>, pageSize: Int, startPostId: Long?, targetUserId: Long?): PageWithCursor<BoardPost> {
+        val pageable = CursorPageable(startPostId, pageSize)
+        val page = fetchPostBasedOnTageExistence(tags, startPostId, targetUserId, pageable)
+
+        val boardPosts = page.content
+        val nextCursorId = if (boardPosts.size < pageSize) null else page.lastOrNull()?.id
+
+        return PageWithCursor(
+            data = boardPosts.map { boardPost ->
+                val userId = boardPost.user?.id
+                userId?.let { BoardPost.from(boardPost, memberDomainService.getMemberProfile(it)) } ?: throw MemberNotFoundException()
+            },
+            nextCursorId = nextCursorId
+        )
```
이렇게 작성하기만 하면 JPA가 알아서 cursor Pagination을 구현해주나요?
구현하신 CursorPageable 객체를 보면 cursorId는 pirvate로 선언되어 있던데 문제 없이 동작하나요?

JinDDung2 replied:
JPA에게 pageable 인터페이스를 건네주면 페이징 처리를 해줘요. 기본적으로 오프셋 기반이어서 CursorPageable을 만들어 사용했어요. 생성할 때 값을 넣어주어서 내부 필드들은 캡슐화를 위해 private 접근 제어자를 사용했어요.

RokwonK replied:
CursorPagable 이란 클래스를 따로 만드신 이유가 있을까요?
cursorId란 필드를 내부적으로 가지고 있긴한데 사용하지 않는 것 같고 일반적인 Pageable와 똑같이 동작하는 것 같아서요.

CursorPagable를 만드신 이유가 id 기반으로 페이징 시키고자 한다면, 작성하신 BoardPostRepository의 쿼리 메서드들에서 CursorPagable의 cursorId를 이용해야 할텐데 그런 부분이 보이지 않는 것 같습니다.

JinDDung2 replied:
처음에 구상할 땐 서버에서 시작점을 알아내 cursorId로 넣어주는 것을 생각했는데, 클라이언트에서 데이터를 건네줄 때 시작점(startId)까지 알려주니 cursorId를 사용하지 않고 startPostId(가독성 위해 구체적인 변수로 수정)만 사용하네요. cursorId필드는 사용하지 않으니 지우는게 나을 것 같아요.

RokwonK commented on adevspoon-domain/src/main/kotlin/com/adevspoon/domain/board/service/BoardPostDomainService.kt:
모든 Post에서는 내가 좋아요 눌렀는지(isLiked)를 확인해야하는데 해당 로직이 안보이는 것 같습니다!

JinDDung2 replied:
그 부분에 대해 상의 할게 있어서 아직 추가를 안했어요. 상의 후에 추가할게요.

RokwonK reviewed (commented):
고생하셨습니다! 코멘트 확인해주세요!

JinDDung2 reviewed (commented):
재리뷰 부탁드립니다.

JinDDung2 committed:
🎨 Transactional 명시적 표기

JinDDung2 committed:
:art: 중복된 정렬 조건을 하나로 합쳐 코드 개선

JinDDung2 committed:
:art: 명세에 따른 변수명 수정

RokwonK commented on adevspoon-domain/src/main/kotlin/com/adevspoon/domain/board/service/BoardPostDomainService.kt:
```diff
@@ -28,12 +33,47 @@ class BoardPostDomainService(
         val boardPost = boardPostRepository.findByIdOrNull(postId) ?: throw BoardPostNotFoundException()
         boardPost.increaseViewCount()
         boardPostRepository.save(boardPost)
+
+        val userId = boardPost.user?.id
+        userId?.let { BoardPost.from(boardPost, memberDomainService.getMemberProfile(it)) } ?: throw MemberNotFoundException()
+
         val memberProfile = memberDomainService.getMemberProfile(userId)
         return BoardPost.from(boardPost, memberProfile)
     }
 
```
BoardPostEntity의 user 필드가 nullble로 되어있어서 헷갈렸을 것 같네요. JoinColumn의 정의를 보시면 nullable = false라 되어 있습니다. 때문에 일단 Entity의 user필드를 아예 non-nullable로 수정하면 좋을 것 같아요. 그렇게 되면 41~43번 줄 코드는 필요 없어질 것 같습니다.

RokwonK commented on adevspoon-domain/src/main/kotlin/com/adevspoon/domain/board/service/BoardPostDomainService.kt:
```diff
@@ -28,12 +33,47 @@ class BoardPostDomainService(
+        val nextCursorId = if (boardPosts.size < pageSize) null else page.lastOrNull()?.id
+
+        return PageWithCursor(
+            data = boardPosts.map { boardPost ->
+                val userId = boardPost.user?.id
+                userId?.let { BoardPost.from(boardPost, memberDomainService.getMemberProfile(it)) } ?: throw MemberNotFoundException()
+            },
```
이 부분도 BoardPostEntity의 user를 non-nullable로 바꾸면 더 가독성 좋은 코드가 될 듯 합니다.

RokwonK reviewed (commented):
Comment 남겼습니다!

JinDDung2 committed:
:sparkles: 로그인 유저가 해당 게시글 좋아요 했는지 확인 여부 추가
- BoardPostEntity, LikeEntity 에서 nullable = false 인 필드는 non-null
- CursorPageable 에서 사용하지 않는 cursorId 삭제

JinDDung2 reviewed (commented):
유저가 해당 게시글 좋아요 눌렀는지 여부, 요청한 리팩토링 진행했어요! 검토부탁해요.

RokwonK commented on adevspoon-domain/src/main/kotlin/com/adevspoon/domain/board/service/BoardPostDomainService.kt:
```diff
@@ -7,33 +7,81 @@ import com.adevspoon.domain.board.exception.BoardTageNotFoundException
+
         val savedBoardPost = boardPostRepository.save(boardPost)
-        return BoardPost.from(savedBoardPost, memberProfile)
+        val isUserLikedBoardPost = likeDomainService.isUserLikedPost(user.id, boardPost.id)
+        return BoardPost.from(savedBoardPost, memberProfile, isUserLikedBoardPost)
```
Board를 만들 당시니 isLike는 무조건 false입니다. 따라서, 해당 로직은 필요없을 듯 합니다!

RokwonK commented on adevspoon-domain/src/main/kotlin/com/adevspoon/domain/board/service/BoardPostDomainService.kt:
```diff
@@ -7,33 +7,81 @@ import com.adevspoon.domain.board.exception.BoardTageNotFoundException
-        val memberProfile = memberDomainService.getMemberProfile(userId)
-        return BoardPost.from(boardPost, memberProfile)
+
+        val user = boardPost.user
+        val isUserLikedBoardPost = likeDomainService.isUserLikedPost(user.id, boardPost.id)
+
```
board 작성자의 like여부가 아닌 API 호출자의 like 여부가 요구사항입니다!

RokwonK commented on adevspoon-domain/src/main/kotlin/com/adevspoon/domain/board/service/BoardPostDomainService.kt:
```diff
@@ -7,33 +7,81 @@ import com.adevspoon.domain.board.exception.BoardTageNotFoundException
+        return BoardPost.from(boardPost, memberProfile, isUserLikedBoardPost)
     }
 
+    @Transactional(readOnly = true)
+    fun getBoardPostsByTags(tags: List<Int>, pageSize: Int, startPostId: Long?, targetUserId: Long?): PageWithCursor<BoardPost> {
+        val pageable = CursorPageable(pageSize)
+        val page = fetchPostBasedOnTageExistence(tags, startPostId, targetUserId, pageable)
+
+        val boardPosts = page.content
+        val nextCursorId = if (boardPosts.size < pageSize) null else page.lastOrNull()?.id
+
+
+
+
+        return PageWithCursor(
+            data = boardPosts.map { boardPost ->
+                val user = boardPost.user
+                val isUserLikedBoardPost = likeDomainService.isUserLikedPost(user.id, boardPost.id)
+                BoardPost.from(boardPost, memberDomainService.getMemberProfile(user.id), isUserLikedBoardPost)
+            },
+            nextCursorId = nextCursorId
+        )
+    }
```
마찬가지로 API를 호출한 유저의 like여부를 반환해야합니다.
또한 위 방식대로라면 N+1(Board List 호출 + 각 Board마다 like 여부호출) 문제가 발생해요!

RokwonK commented on adevspoon-domain/src/main/kotlin/com/adevspoon/domain/board/service/BoardPostDomainService.kt:
```diff
@@ -7,33 +7,81 @@ import com.adevspoon.domain.board.exception.BoardTageNotFoundException
 
+    @Transactional(readOnly = true)
+    fun getBoardPostsByTags(tags: List<Int>, pageSize: Int, startPostId: Long?, targetUserId: Long?): PageWithCursor<BoardPost> {
+        val pageable = CursorPageable(pageSize)
```
pageSize만 넘기는 것이라면 CursorPageable의 존재 의미가 사라지지 않나요?

JinDDung2 replied:
CursorPageable에는 pageSize, Sort가 있어요. Sort의 디폴트를 id 내림차순으로 만들어둬서 jpql을 사용할 때 내림차순을 반복적으로 적지 않아도 되요.

RokwonK reviewed (commented):
고생하셨습니다. like 여부와 관련되어서 코멘트 한 번 확인해주세요!

JinDDung2 committed:
:zap: 로그인 유저가 해당 게시글 좋아요 했는지 확인 기능 수정 및 N+1 문제 해결

RokwonK reviewed (approved):
고생하셨어요!! LGTM~
